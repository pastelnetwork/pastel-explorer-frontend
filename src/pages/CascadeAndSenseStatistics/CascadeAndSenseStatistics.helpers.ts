import format from 'date-fns/format';

import { TAverageRarenessScoreData, TCascadeAndSenseData } from '@utils/types/IStatistics';

export const transformAverageRarenessScoreOfNFTsOnSenseChartData = (
  data: TAverageRarenessScoreData[] | null,
  range = 1,
) => {
  const dataX: string[] = [];
  const dataY: number[] = [];
  const dataY1: number[] = [];
  if (data?.length) {
    data.forEach(item => {
      dataX.push(format(item.timestamp, 'MM/dd/yyyy hh:mm aa'));
      dataY.push(item.highest ? item.highest / range : 0);
      dataY1.push(item.average ? item.average / range : 0);
    });
  }

  return {
    dataX,
    dataY,
    dataY1,
  };
};

export const transformChartData = (data: TCascadeAndSenseData[] | null, range = 1) => {
  const dataX: string[] = [];
  const dataY: number[] = [];
  if (data?.length) {
    data.forEach(item => {
      dataX.push(format(item.timestamp, 'MM/dd/yyyy hh:mm aa'));
      dataY.push(item.value ? item.value / range : 0);
    });
  }

  return {
    dataX,
    dataY,
  };
};
