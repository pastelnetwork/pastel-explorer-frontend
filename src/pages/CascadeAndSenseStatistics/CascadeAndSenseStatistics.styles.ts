import styled from 'styled-components';
import { Typography as MuiTypography } from '@material-ui/core';
import { rgba } from 'polished';

export const Wrapper = styled.div`
  width: 100%;

  .cascade-sense-statistics {
    .MuiSelect-select {
      width: 60px;
      border: 0;
      padding: 0 20px 0;
    }

    .MuiSelect-icon {
      right: -6px;
    }
  }

  .cascade-sense-card {
    width: calc(33% - 4px);
    margin-right: 12px;

    &:nth-child(3),
    &:nth-child(6) {
      margin-right: 0;
    }

    .echarts-for-react {
      height: 150px !important;
    }

    ${props => props.theme.breakpoints.down('lg')} {
      margin-right: 10px;

      &:nth-child(3),
      &:nth-child(6) {
        margin-right: 0;
      }
    }

    ${props => props.theme.breakpoints.down('md')} {
      width: calc(50% - 6px);
      margin-right: 12px;

      &:nth-child(3) {
        margin-right: 12px;
      }

      &:nth-child(2),
      &:nth-child(4),
      &:nth-child(6) {
        margin-right: 0;
      }
    }

    ${props => props.theme.breakpoints.down(767)} {
      width: 100%;
      margin-right: 0;
      margin-left: 0;
    }

    ${props => props.theme.breakpoints.down(321)} {
      &.total-size-of-data-stored,
      &.average-size-of-nft-stored,
      &.total-fingerprints,
      &.average-rareness-score {
        .cascade-sense-statistics {
          min-height: 35px;
          align-items: flex-start;
        }
      }
    }
  }
`;

export const HeaderWrapper = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 20px;
  padding: 14px 16px;
  background: ${props => props.theme.card.titleColor};
  border-radius: 10px;
  overflow: hidden;

  ${props => props.theme.breakpoints.down(500)} {
    flex-direction: column;
    justify-content: flex-start;
    align-items: flex-start;
  }
`;

export const HeaderTitle = styled.div`
  display: flex;
  align-items: center;
  border: 0;
  background: transparent;
  color: ${props => props.theme.palette.text.primary};
  font-size: 16px;
  font-weight: 700;
  line-height: 1.3;
  transition: all 0.5s ease;
`;

export const ChartWrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  width: 100%;
  padding: 0;
`;

export const ContentWrapper = styled.div`
  width: calc(50% - 10px);
  margin-right: 20px;
  margin-bottom: 20px;
  padding: 0;
  border-radius: 10px;
  box-shadow: 0px 5px 6px rgb(16 16 16 / 6%);

  &.full {
    width: 100%;
    margin-right: 0;
  }

  &:nth-child(even) {
    margin-right: 0;
  }

  ${props => props.theme.breakpoints.down(750)} {
    width: 100%;
    margin-right: 0;
    padding: 0;
  },
`;

export const Percentage = styled.div`
  font-size: 0.8rem;
  color: ${props => props.theme.sidebar.menu.default};

  span {
    display: inline-block;
    margin-top: ${props => props.theme.spacing(1)}px;
    color: ${props => props.theme.palette.text.secondary};
    padding: 2px 6px;
    border-radius: 3px;
    font-weight: 500;
    font-size: 16px;
  }
`;

export const PercentageValue = styled(MuiTypography)<{
  percentagecolor: string;
  mb: number;
}>`
  font-size: 0.8rem;
  color: ${props => props.theme.sidebar.menu.default};

  span {
    display: inline-block;
    margin-top: ${props => props.theme.spacing(1)}px;
    color: ${props => props.theme.palette.text.secondary};
    background: ${props => rgba(props.percentagecolor, 0.2)};
    padding: 2px 6px;
    border-radius: 3px;
    font-weight: 500;
    font-size: 16px;
  }
`;

export const ChartContentWrapper = styled.div`
  position: relative;
  width: 100%;
  height: 170px;
  margin-top: ${props => props.theme.spacing(2.5)}px;
  border-radius: 6px;
  overflow: hidden;
`;
