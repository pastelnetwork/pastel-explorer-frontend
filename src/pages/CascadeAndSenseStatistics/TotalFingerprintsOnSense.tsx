import { useState, ChangeEvent } from 'react';
import { Skeleton } from '@material-ui/lab';

import { PeriodTypes, generatePeriodToDropdownOptions } from '@utils/helpers/statisticsLib';
import { formatNumber } from '@utils/helpers/formatNumbers/formatNumbers';
import { periods } from '@utils/constants/statistics';
import { LineChart } from '@components/Summary/LineChart';
import { Dropdown } from '@components/Dropdown/Dropdown';
import themeVariant from '@theme/variants';
import useTotalFingerprintsOnSense from '@hooks/useTotalFingerprintsOnSense';
import { translate } from '@utils/helpers/i18n';
import * as SummaryStyles from '@components/Summary/Summary.styles';
import * as StatisticsStyles from '@pages/Statistics/Statistics.styles';

import { transformChartData } from './CascadeAndSenseStatistics.helpers';
import * as Styles from './CascadeAndSenseStatistics.styles';

const TotalFingerprintsOnSense: React.FC = () => {
  const [period, setPeriod] = useState<PeriodTypes>(periods[7][2]);
  const { data, isLoading, difference, currentValue } = useTotalFingerprintsOnSense(period);

  const handleDropdownChange = (
    event: ChangeEvent<{
      value: unknown;
    }>,
  ) => {
    if (event.target.value) {
      setPeriod(event.target.value as PeriodTypes);
    }
  };
  const chartData = transformChartData(data);

  return (
    <SummaryStyles.Card className="cascade-sense-card total-fingerprints">
      <SummaryStyles.CardContent>
        <SummaryStyles.ValueWrapper>
          <SummaryStyles.Typography variant="h6">
            {translate('pages.cascadeAndSenseStatistics.totalFingerprintsOnSense')}
          </SummaryStyles.Typography>
          <SummaryStyles.Typography variant="h4">
            <SummaryStyles.Values>
              {isLoading ? (
                <Skeleton animation="wave" variant="text" />
              ) : (
                formatNumber(currentValue)
              )}
            </SummaryStyles.Values>
          </SummaryStyles.Typography>
        </SummaryStyles.ValueWrapper>
        <SummaryStyles.PercentageWrapper>
          <Styles.Percentage>
            <Dropdown
              value={period}
              onChange={handleDropdownChange}
              options={generatePeriodToDropdownOptions(periods[7])}
              classNameWrapper="cascade-sense-statistics"
            />
            <br />
            <Styles.PercentageValue
              variant="subtitle2"
              mb={4}
              color="textSecondary"
              noWrap
              percentagecolor={`${
                difference > 0 ? themeVariant.custom.green.success : themeVariant.custom.red.error
              }`}
            >
              <span>
                {`${difference > 0 ? '+' : ''}`}
                {difference}%&nbsp;
                {difference > 0 ? (
                  <svg
                    width="10"
                    height="10"
                    viewBox="0 0 10 10"
                    fill="none"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      fillRule="evenodd"
                      clipRule="evenodd"
                      d="M4.23381 2.93331L1.08356 6.08356L4.37114e-07 5L5 4.37114e-07L10 5L8.91644 6.08356L5.76619 2.93331L5.76619 10L4.23381 10L4.23381 2.93331Z"
                      fill="#00D097"
                    />
                  </svg>
                ) : (
                  <svg
                    width="10"
                    height="10"
                    viewBox="0 0 10 10"
                    fill="none"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      fillRule="evenodd"
                      clipRule="evenodd"
                      d="M5.76619 7.06669L8.91644 3.91644L10 5L5 10L-2.18557e-07 5L1.08356 3.91644L4.23381 7.06669L4.23381 -2.52048e-07L5.76619 -1.85066e-07L5.76619 7.06669Z"
                      fill="#FF754C"
                    />
                  </svg>
                )}
              </span>
            </Styles.PercentageValue>
          </Styles.Percentage>
        </SummaryStyles.PercentageWrapper>
      </SummaryStyles.CardContent>
      <Styles.ChartContentWrapper>
        {isLoading ? (
          <StatisticsStyles.Loader>
            <Skeleton animation="wave" variant="rect" height={170} width="100%" />
            <StatisticsStyles.LoadingText>
              {translate('common.loadingData')}
            </StatisticsStyles.LoadingText>
          </StatisticsStyles.Loader>
        ) : (
          <LineChart
            chartName="totalFingerprintsOnSense"
            dataX={chartData?.dataX}
            dataY={chartData?.dataY}
            offset={1}
            disableClick
          />
        )}
      </Styles.ChartContentWrapper>
    </SummaryStyles.Card>
  );
};

export default TotalFingerprintsOnSense;
