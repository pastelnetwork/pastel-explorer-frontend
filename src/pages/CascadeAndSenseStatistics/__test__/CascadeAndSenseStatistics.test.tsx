import { shallow } from 'enzyme';

import 'jest-styled-components';

import i18next from '../../../utils/helpers/i18n';
import CascadeAndSenseStatistics from '../CascadeAndSenseStatistics';

jest.mock('i18next-http-backend');
jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () =>
          new Promise(() => {
            // noop
          }),
      },
    };
  },
  initReactI18next: {
    type: '3rdParty',
    init: () => {
      // noop
    },
  },
}));
i18next.t = jest.fn().mockImplementation((...arg) => {
  return arg[0];
});

describe('pages/CascadeAndSenseStatistics', () => {
  const wrapper = shallow(<CascadeAndSenseStatistics />);

  test('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
