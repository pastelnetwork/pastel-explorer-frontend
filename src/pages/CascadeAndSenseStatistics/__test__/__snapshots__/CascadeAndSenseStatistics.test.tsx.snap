// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/CascadeAndSenseStatistics renders correctly 1`] = `ShallowWrapper {}`;
